prefixdb_test.go

```diff
@@ -20,6 +25,86 @@ func mockDBWithStuff(t *testing.T) DB {
 	return db
 }
 
+func taskKey(i, k int) []byte {
+	return []byte(fmt.Sprintf("task-%d-key-%d", i, k))
+}
+
+func randomValue() []byte {
+	b := make([]byte, 16)
+	_, err := rand.Read(b)
+
+	if err != nil {
+		return nil
+	}
+
+	return b
+}
+
+// Run generates concurrent reads and writes to db so the race detector can
+// verify concurrent operations are properly synchronized.
+// The contents of db are garbage after Run returns.
+func Run(t *testing.T) {
+
+	db := mockDBWithStuff(t)
```

Comment:
It won't help if we use a mock database here—this needs to exercise the real database implementation. I suggest making this a parameter, e.g.,

```go
func Run(t *testing.T, db DB) {
  // ...
}
```

Then we can use it to write more specific tests against the individual implementations, e.g.,

```go
func TestBadger(t *testing.T) {
   db := newTestBadgerDB(t) // or whatever we do to set one up
   t.Run("Badger", func(t *testing.T) { Run(t, db) })
}
```

or words to that effect. Exercising the MemDB is fine, but doesn't tell us much about the concurrency safety of the "real" backends.

Reply:
gotcha!

OK, will address this tomorrow, and thanks!

Reply:
should we test prefix_db with many db implementations, like some kind of table-driven test

Reply:
I do think that is what he means <3